apps/web/pages/api/cron/syncAppMeta.ts

```diff
@@ -0,0 +1,58 @@
+import type { NextApiRequest, NextApiResponse } from "next";
+
+import { getAppWithMetadata } from "@calcom/app-store/_appRegistry";
+import { prisma } from "@calcom/prisma";
+import type { AppCategories, Prisma } from "@calcom/prisma/client";
+
+/**
+ * syncAppMeta makes sure any app metadata that has been replicated into the database
+ * remains synchronised with any changes made to the app config files.
+ */
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  const apiKey = req.headers.authorization || req.query.apiKey;
+  if (process.env.CRON_API_KEY !== apiKey) {
+    res.status(401).json({ message: "Not authenticated" });
+    return;
+  }
+  if (req.method !== "POST") {
+    res.status(405).json({ message: "Invalid method" });
+    return;
+  }
+
+  console.log(`🧐 Checking DB apps are in-sync with app metadata`);
+
+  const dbApps = await prisma.app.findMany();
+
+  for await (const dbApp of dbApps) {
+    const app = await getAppWithMetadata(dbApp);
+    const updates: Prisma.AppUpdateManyMutationInput = {};
+
+    if (!app) {
+      console.error(`💀 App ${dbApp.slug} (${dbApp.dirName}) no longer exists.`);
+      continue;
+    }
+
+    if (
+      dbApp.categories.length !== app.categories.length ||
+      !dbApp.categories.every((category) => app.categories.includes(category))
+    ) {
+      updates["categories"] = app.categories as AppCategories[];
```

Comment:
It is updated only when there is a category for an App in DB that doesn't exist now.
We should also do the reverse thing, where there is an category in file for an app but it doesn't exist in DB.

Reply:
I think I understand what you're referring to here, and I think that's already happening – I could have dropped a comment to make it a little clearer what these conditions are doing:

By combining the length check and "every" check, these two conditions are making sure the two lists are identical in content. Let me re-validate my logic by typing out the various scenarios:

1. If we have a category removed in the app metadata, then `dbApp.categories.length` will be greater than `app.categories.length`. The first condition will therefore pass, and the categories will be updated.

2. If we have a new category added in the app metadata (the case you're highlighting), the same will be true – just in the other direction.

3. If we had both a category removal and addition then the length of those lists will be the same. We'll then step through each item in _one_ of the lists (doesn't matter which) and make sure they're also present in the other list. If any of them are different, we'll update the categories.

    (Typing this out, it may actually be _technically_ more appropriate and efficient to use `dbApp.categories.some` rather than `!dbApp.categories.every` – although that difference in efficiency won't be noticeable at the scale we're dealing with)

4. Finally, in the case that there are no changes to the categories, the list lengths will be the same, and the second condition will pass as every item in one list is also present in the other list.

Let me know if that helps. I'd be happy to jump on a call and chat through further if it's still unclear, or if I've completely misinterpreted your feedback!

Reply:
Ah my bad. Thanks !!